runtime/src/builtins.rs

```diff
@@ -51,11 +51,61 @@ macro_rules! with_program_logging {
     };
 }
 
-#[derive(AbiExample, Debug, Clone)]
-pub enum ActivationType {
-    NewProgram,
-    NewVersion,
-    RemoveProgram,
+/// State transition enum used for adding and removing builtin programs through
+/// feature activations.
+#[derive(Debug, Clone)]
+pub enum BuiltinFeatureTransition {
+    /// Add a builtin program if a feature is activated.
+    Add {
+        builtin: Builtin,
+        feature_id: Pubkey,
+    },
+    /// Remove a builtin program if a feature is activated or
+    /// retain a previously added builtin.
+    RemoveOrRetain {
+        previous_builtin: Builtin,
+        removal_feature_id: Pubkey,
+    },
```

Comment:
also, I'm omitting `Replace` as it's really useless now. That's because builtins now has access to feature_set. `Replace` was relic before the widespread use of feature activation mechanism...

Reply:
lastly, i wonder when `Remove` would be needed in the (distant) future. so, I'm just saying leaving a comment to work as a pointer to this pr.

Reply:
Yes, that's exactly why I removed Replace. Relic of the past and not useful anymore.

Reply:
Great yeah I will add a comment, I agree it probably won't be needed in the future